### ISSUE TITLE
get rid of branches that deploy to multiple scriptworkers

### ISSUE THREAD

**Issue by bhearsum** [mozilla-releng/scriptworker-scripts#359]
We have `production` (and I think `dev`) that deploy to all the scriptworkers. IMO these are a big footgun because it makes it difficult to know what exactly is deployed somewhere. On multiple occasions I have accidentally reverted something because it was deployed through, eg: `production` instead of `production-balrogscript`.

@escapewindow - curious your thoughts here.

**Comment by escapewindow** [comment]
Tl;dr: I'm curious what were the root causes of the accidental reversions, and I think the future is relpro.

> We have `production` (and I think `dev`) that deploy to all the scriptworkers.

Correct.

> IMO these are a big footgun because it makes it difficult to know what exactly is deployed somewhere. On multiple occasions I have accidentally reverted something because it was deployed through, eg: `production` instead of `production-balrogscript`.

Hm. The question that comes up here is "why?" Did we have something pushed to `production` or `production-balrogscript` that wasn't merged to `master` (this is not best practice)? Or were you pushing an outdated revision to one of those branches, in which case, why weren't you pushing the latest master revision?

> @escapewindow - curious your thoughts here.

I use both. Generally `production` whenever I need to deploy something to all scriptworkers, and `production-*script` when I want to deploy something individually. Pushing to all of the latter is a pain when I want to roll out a scriptworker fix everywhere at once, especially if it's time sensitive. Pushing to the former is a pain if I just want to roll out something in one pool and I don't want to redeploy everything.

I think the future is relpro. We'll have promotable docker images built on merge to `master`, but they won't go anywhere automatically. We can choose to promote them all to production through shipit, or just promote a subset. Ideally we'd have some sort of dashboard that shows us what the latest deployed images are, and ideally some sort of history so we can roll back easily: just re-promote the previous image(s), rather than have to rebuild and deal with docker image builds being non-deterministic. In fact, if we had such a dashboard now, that would help, but I tend to try to figure out what the latest deploy(s) are through slack scrollback and branch history. A git pushlog would help.

**Comment by bhearsum** [comment]
> Tl;dr: I'm curious what were the root causes of the accidental reversions, and I think the future is relpro.
> 
> > We have `production` (and I think `dev`) that deploy to all the scriptworkers.
> 
> Correct.
> 
> > IMO these are a big footgun because it makes it difficult to know what exactly is deployed somewhere. On multiple occasions I have accidentally reverted something because it was deployed through, eg: `production` instead of `production-balrogscript`.
> 
> Hm. The question that comes up here is "why?" Did we have something pushed to `production` or `production-balrogscript` that wasn't merged to `master` (this is not best practice)? Or were you pushing an outdated revision to one of those branches, in which case, why weren't you pushing the latest master revision?

Part of the trouble here was that I was doing a push with no new changes to this repo - I was just trying to trigger a deploy to pick up cloudops changes. And in this case, I was just trying to deploy a single scriptworker.

IIRC, I was looking at a `production-X` branch and saw that it didn't have the scriptworker 38.x upgrade. I was uncertain as to whether that was safe to take, and it wasn't clear to me what exactly we were running currently. It _seemed_ like the safer route was just to push a no-op change to that branch.

My memory is telling me that the `production-X` branch looked newer than `production` at the time, but I can't confirm that.

> > @escapewindow - curious your thoughts here.
> 
> I use both. Generally `production` whenever I need to deploy something to all scriptworkers, and `production-*script` when I want to deploy something individually. Pushing to all of the latter is a pain when I want to roll out a scriptworker fix everywhere at once, especially if it's time sensitive. Pushing to the former is a pain if I just want to roll out something in one pool and I don't want to redeploy everything.
> 
> I think the future is relpro. We'll have promotable docker images built on merge to `master`, but they won't go anywhere automatically. We can choose to promote them all to production through shipit, or just promote a subset. Ideally we'd have some sort of dashboard that shows us what the latest deployed images are, and ideally some sort of history so we can roll back easily: just re-promote the previous image(s), rather than have to rebuild and deal with docker image builds being non-deterministic. In fact, if we had such a dashboard now, that would help, but I tend to try to figure out what the latest deploy(s) are through slack scrollback and branch history. A git pushlog would help.

That would be lovely. I think these problems would be largely mitigated if it was easier to be certain about what _exactly_ is deployed at the moment. This is easier for things like Ship It and Balrog, where we have `/__version__` endpoints with revision in them, but tougher for things like scriptworkers where there's no way to query them.

**Comment by bhearsum** [comment]
(feel free to close this if you want)

**Comment by escapewindow** [comment]
> Part of the trouble here was that I was doing a push with no new changes to this repo - I was just trying to trigger a deploy to pick up cloudops changes. And in this case, I was just trying to deploy a single scriptworker.

Ah, ok.

> IIRC, I was looking at a `production-X` branch and saw that it didn't have the scriptworker 38.x upgrade. I was uncertain as to whether that was safe to take, and it wasn't clear to me what exactly we were running currently. It _seemed_ like the safer route was just to push a no-op change to that branch.
> 
> My memory is telling me that the `production-X` branch looked newer than `production` at the time, but I can't confirm that.

I tend to look for the most recent k8s-image task off either `production` or `production-*script` for that pool, and then either force rerun it (pre-deadline) or retrigger it (post-deadline), which should work without pushing dummy changes. Agreed it can be tough to find that task, and agreed that having both branches is a pain when you just want to deploy one thing.

> > I think the future is relpro. We'll have promotable docker images built on merge to `master`, but they won't go anywhere automatically. We can choose to promote them all to production through shipit, or just promote a subset. Ideally we'd have some sort of dashboard that shows us what the latest deployed images are, and ideally some sort of history so we can roll back easily: just re-promote the previous image(s), rather than have to rebuild and deal with docker image builds being non-deterministic. In fact, if we had such a dashboard now, that would help, but I tend to try to figure out what the latest deploy(s) are through slack scrollback and branch history. A git pushlog would help.
> 
> That would be lovely. I think these problems would be largely mitigated if it was easier to be certain about what _exactly_ is deployed at the moment. This is easier for things like Ship It and Balrog, where we have `/__version__` endpoints with revision in them, but tougher for things like scriptworkers where there's no way to query them.

Yeah. I wonder if we can update the cloudops-jenkins task to update some db somewhere. Or the k8s-image task could as well, if the [push-to-dockerhub flag is true](https://github.com/mozilla-releng/scriptworker-scripts/blob/master/taskcluster/docker/k8s-image/build_and_push.sh#L19). I suppose we can weed through the [tags on dockerhub](https://hub.docker.com/r/mozilla/releng-balrogscript/tags?page=1&ordering=last_updated) but they're not so friendly either.

**Comment by escapewindow** [comment]
> (feel free to close this if you want)

I'd like to improve this in some way, even if we don't get rid of the `production` branch.
